src/ESPAsyncTCP.cpp

```diff
@@ -265,6 +265,9 @@ bool AsyncClient::connect(IPAddress ip, uint16_t port){
   tcp_arg(pcb, this);
   tcp_err(pcb, &_s_error);
   size_t err = tcp_connect(pcb, &addr, port,(tcp_connected_fn)&_s_connected);
+  if (err != ERR_OK){
+    tcp_close(pcb);
```

Comment:
Do you mean that `pcb` object leaks?
Does the err callback gets called at some point? (during tcp_connect or some time after)? Does the timeout mechanism eventually (~15min) purge it from memory?

Should we call `::_close()`, not lwip api? tcp_close also returns ERR, which is handled there

Reply:
In my test: if wifi is not connected, the tcp_connect will report "Routing problem", and _s_connected seem to not called back, and member variable _pcb is not assigned, so ::_close() cannot free the pcb object.

In my project: https://github.com/xfan1024/HiButton
if do not apply the patch, the program will crashed on boot

Reply:
My bad, I missed the `_pcb` underscore.
It does sound like a good practice, lwip-contrib and mqtt & lwiperf apps in the lwip tree itself also explicitly close the pcb after failed connection. What I meant to add is should we use abort too?
```diff
diff --git a/src/ESPAsyncTCP.cpp b/src/ESPAsyncTCP.cpp
index 7c3fd97..330b5d7 100644
--- a/src/ESPAsyncTCP.cpp
+++ b/src/ESPAsyncTCP.cpp
@@ -130,8 +130,13 @@ bool AsyncClient::connect(IPAddress ip, uint16_t port){
 #endif
   tcp_arg(pcb, this);
   tcp_err(pcb, &_s_error);
-  tcp_connect(pcb, &addr, port,(tcp_connected_fn)&_s_connected);
-  return true;
+  err_t err = tcp_connect(pcb, &addr, port,(tcp_connected_fn)&_s_connected);
+  if (err != ERR_OK) {
+    if(tcp_close(pcb) != ERR_OK) {
+        tcp_abort(pcb);
+    }
+  }
+  return (err == ERR_OK);
 }
```

I'm not sure if lwip won't eventually delete the pcb in question though, but the code example in question calls connect() every loop(), so there is no time to check that and memory is already gone in less than a second.

Reply:
``` c
  /* check if we have a route to the remote host */
  if (ip_addr_isany(&(pcb->local_ip))) {
    /* no local IP address set, yet. */
    struct netif *netif = ip_route(&(pcb->remote_ip));
    if (netif == NULL) {
      /* Don't even try to send a SYN packet if we have no route
         since that will fail. */
      return ERR_RTE;
    }
    /* Use the netif's IP address as local address. */
    ip_addr_copy(pcb->local_ip, netif->ip_addr);
  }

```
I am not good at LwIP. In my opinion if route check fail LwIP will not do any thing else for the pcb object, so the pcb object can only released by caller. Calling tcp_abort() is not neccessary, in lwip-contrib/apps/smtp/smtp.c

```c
static err_t
smtp_send_mail_alloced(struct smtp_session *s)
{
  ...
    err = tcp_connect(pcb, &addr, smtp_server_port, smtp_tcp_connected);
    if (err != ERR_OK) {
      LWIP_DEBUGF(SMTP_DEBUG_WARN_STATE, ("tcp_connect failed: %d\n", (int)err));
      goto deallocate_and_leave;
    }
  ...
deallocate_and_leave:
  if (pcb != NULL) {
    tcp_arg(pcb, NULL);
    tcp_close(pcb);
  }
  ...
}
```

the example only call tcp_close(pcb)